cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -189,6 +194,73 @@ Status ConstBoolExec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
   return Status::OK();
 }
 
+struct NonZeroVisitor {
+  UInt64Builder* builder;
+  const ArrayData& array;
+
+  NonZeroVisitor(UInt64Builder* builder, const ArrayData& array)
+      : builder(builder), array(array) {}
+
+  Status Visit(const DataType& type) { return Status::NotImplemented(type.ToString()); }
+
+  template <typename Type>
+  enable_if_t<is_primitive_ctype<Type>::value, Status> Visit(const Type&) {
+    using T = typename GetViewType<Type>::T;
+    uint32_t index = 0;
+
+    return VisitArrayDataInline<Type>(
+        this->array,
+        [&](T v) {
+          if (v) {
+            this->builder->UnsafeAppend(index);
+          }
+          ++index;
+          return Status::OK();
+        },
+        [&]() {
+          ++index;
+          return Status::OK();
+        });
+  }
+};
+
+Status IndicesNonZeroExec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  std::shared_ptr<ArrayData> array = batch[0].array();
+  UInt64Builder builder;
+
+  RETURN_NOT_OK(builder.Reserve(array->length));
+  NonZeroVisitor visitor(&builder, *array.get());
+  RETURN_NOT_OK(VisitTypeInline(*(array->type), &visitor));
+
+  std::shared_ptr<ArrayData> out_data;
+  RETURN_NOT_OK(builder.FinishInternal(&out_data));
+  out->value = std::move(out_data);
+  return Status::OK();
+}
+
+std::shared_ptr<ScalarFunction> MakeIndicesNonZeroFunction(std::string name,
```

Comment:
So, this is not a scalar function as the output doesn't have the same length as the input.
(a scalar function is a function where each value in the output can be derived independently from the corresponding value in the input)

This should be registered as a vector function.

Reply:
Oh thanks for the clarification, I thought that a function could be considered scalar as far as it only cared about one element of the array at a time, I'll change this to be a `VectorFunction`

Reply:
At this point, I think the code should also be moved outside of `scalar_validity.cc`, and into `vector_?.cc`, is there any suggestion about where it would fit best?

Reply:
:+1: done